### PR TITLE
feat(connector): [PPRO] add Afterpay and UPI Intent/QR support

### PR DESCRIPTION
**Pull request by ankush370**
---
  Changes

  - Add Afterpay (AfterpayClearpay) PayLater support, requiring consumer email and billing country on the charge request
  - Add UPI Intent (AppIntent) and UPI QR (ScanCode) authentication flows
  - Consolidate auth-method redirect selection into build_auth_redirect based on Payment Method
  - Map PPRO refund-pending webhooks to a new RefundProcessing event (proto WEBHOOK_REFUND_PROCESSING, EventType, IncomingWebhookEvent)
  - Register UpiQr and AfterpayClearpay in PPRO supported payment methods
  - Surface merchant_transaction_id on the refund response/webhook, populated for PPRO from merchant_payment_charge_reference, and send merchant_refund_reference in the refund request

  Description

  Adds Afterpay (BNPL) support and UPI Intent / UPI QR authentication flows to the PPRO connector, introduces a RefundProcessing webhook event so refund-pending notifications are surfaced correctly, and adds a merchant transaction reference to refund responses.

  - Afterpay (PayLater): maps AfterpayClearpay → "AFTERPAY", registers it in PPRO's supported payment methods (refunds supported, mandates not), and requires consumer email + billing country on the charge request.
  - UPI: maps UpiIntent → AppIntent (mobile app-switch) and UpiQr → ScanCode (scan-to-pay) authentication types; registers UpiQr as a supported method.
  - Auth redirect handling: consolidates the duplicated redirect-selection loops into a single build_auth_redirect helper based on the payment method passed in request.
  - Webhooks: adds WEBHOOK_REFUND_PROCESSING / RefundProcessing and maps PPRO's PaymentChargeRefundPending (and the new created / authentication-pending / provider-confirmation-pending charge events) to the correct processing / action-required states.
- Webhooks: adds WEBHOOK_REFUND_PROCESSING / RefundProcessing and maps PPRO's PaymentChargeRefundPending (and the new created / authentication-pending / provider-confirmation-pending charge events) to the correct processing / action-required states.
- Refund references: adds merchant_transaction_id to RefundWebhookDetailsResponse and the RefundResponse proto so the merchant's original transaction reference is surfaced on refunds. PPRO populates it from merchant_payment_charge_reference and sends merchant_refund_reference in the refund request; all other connectors set it to None.

Motivation and Context

PPRO previously only supported the Redirect authentication flow and had no Afterpay support. UPI payments require app-intent (Intent) and QR (scan-code) flows, and Afterpay requires consumer email/country to be sent. PPRO also emits a refund-pending webhook that had no corresponding event type, so it was being returned as EventNotSupported — this PR adds RefundProcessing to represent it accurately. Finally, refund flows now carry the merchant's original transaction reference for easier reconciliation.

Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies application configuration/environment variables

Modified API contract:
crates/types-traits/grpc-api-types/proto/payment.proto
- new WEBHOOK_REFUND_PROCESSING = 42 value in WebhookEventType
- new optional field RefundResponse.merchant_transaction_id = 24

## How did you test it?
Tested via Composite APIs

Request - 
```
{
    "order_tax_amount": null,
    "order_category": null,
    "order_details": [],
    "request_extended_authorization": null,
    "threeds_completion_indicator": null,
    "return_url": "https://webhook.site/468a507c-e60d-44cc-b5c6-07a6e55d35c5",
    "test_mode": true,
    "off_session": null,
    "merchant_account_metadata": null,
    "metadata": "{\"order_id\":\"JUSP1782206838\"}",
    "billing_descriptor": null,
    "capture_method": "AUTOMATIC",
    "statement_descriptor_name": null,
    "webhook_url": "https://webhook.site/468a507c-e60d-44cc-b5c6-07a6e55d35c5",
    "feature_data": null,
    "customer": {
        "last_name": null,
        "name": null,
        "email": null,
        "connector_customer_id": null,
        "phone_number": null,
        "first_name": null
    },
    "browser_info": {
        "screen_height": null,
        "ip_address": null,
        "accept_header": "*/*",
        "accept_language": null,
        "referer": null,
        "java_enabled": null,
        "color_depth": null,
        "os_type": null,
        "device_model": null,
        "screen_width": null,
        "user_agent_string": null,
        "time_zone_offset": null,
        "java_script_enabled": null,
        "language": null,
        "os_version": null
    },
    "description": "",
    "enrolled_for_3ds": null,
    "state": null,
    "session_token": null,
    "request_incremental_authorization": null,
    "payment_method_token": null,
    "amount": {
        "currency": "GBP",
        "minor_amount": 100
    },
    "merchant_order_id": "JUSP1782206838",
    "redirection_response": null,
    "address": {
        "shipping_address": {
            "last_name": null,
            "line2": null,
            "phone_country_code": null,
            "zip_code": null,
            "email": null,
            "state": null,
            "line3": null,
            "phone_number": null,
            "country_alpha2_code": "US",
            "city": null,
            "line1": null,
            "first_name": null
        },
        "billing_address": {
            "last_name": null,
            "line2": null,
            "phone_country_code": null,
            "zip_code": null,
            "email": null,
            "state": null,
            "line3": null,
            "phone_number": null,
            "country_alpha2_code": "GB",
            "city": null,
            "line1": null,
            "first_name": null
        }
    },
    "complete_authorize_url": null,
    "auth_type": "AUTHENTICATION_TYPE_UNSPECIFIED",
    "payment_method": {
        "payment_method": {
            "upi_intent": {}
        }
    },
    "statement_descriptor_suffix": null,
    "shipping_cost": null,
    "merchant_request_id": "5362d1e8-c34f-4b36-8229-4a89b9dcef1b",
    "continue_redirection_url": null,
    "enable_partial_authorization": null,
    "setup_future_usage": null,
    "merchant_transaction_id": "azharamin-JUSP1782206838-1"
}
```
Connector Call - 
```
 {
    "paymentMethod": "UPI",
    "paymentMedium": "ECOMMERCE",
    "merchantPaymentChargeReference": "azharamin-JUSP1782206838-1",
    "paymentDescriptor": "",
    "amount": {
      "currency": "GBP",
      "value": 100
    },
    "consumer": {
      "name": "*** alloc::string::String ***",
      "email": "***************@gmail.com",
      "country": "GB"
    },
    "authenticationSettings": [
      {
        "type": "APP_INTENT"
      }
    ],
    "webhooksUrl": "https://webhook.site/468a507c-e60d-44cc-b5c6-07a6e55d35c5"
  }
```
Response - 
```
{
    "merchant_transaction_id": "azharamin-JUSP1782206838-1",
    "connector_transaction_id": "charge_EQpTlLNhdKX2U2whgso3Q",
    "status": "AUTHENTICATION_PENDING",
    "status_code": 200,
    "response_headers": {},
    "redirection_data": {
      "form_type": {
        "Uri": {
          "uri": "upiGlobal://pay?ver=01&from_app=true&paymentChargeId=charge_EQpTlLNhdKX2U2whgso3Q&amount=100&currency=GBP"
        }
      }
    },
    "state": {
      "connector_customer_id": "cth_rUeHPRtCzTmNdX93"
    },
    "mandate_reference": {
      "mandate_id_type": {
        "ConnectorMandateId": {
          "connector_mandate_id": "instr_UNEjh8hIAWg8JtQ73Aq9l"
        }
      }
    }
  }
```
Refund Pending Webhook - 
```
{
    "source": "https://www.ppro.com",
    "id": "ZLGpp0B6Ln7StbBFJsD2q",
    "type": "PAYMENT_CHARGE_REFUND_PENDING",
    "subject": "charge_5KKLI7IJbcaNAi2oPUm62",
    "time": "2026-06-30T11:05:38.959Z",
    "data": {
        "paymentChargeId": "charge_5KKLI7IJbcaNAi2oPUm62",
        "paymentChargeStatus": "CAPTURED",
        "paymentMethod": "UPI",
        "merchantId": "BARCLAYSGAPITEST",
        "merchantPaymentChargeReference": "Test-order-1",
        "type": "PAYMENT_CHARGE_REFUND_PENDING",
        "refundId": "refund_SoDJwDyhCtW5OmAM9WUzP",
        "amount": {
            "value": 100,
            "currency": "GBP"
        },
        "merchantRefundReference": "ref1782817526"
    },
    "specversion": "1.0.2",
    "datacontenttype": "application/json"
}
```
Response - 
```
{
    "event_type": "WEBHOOK_REFUND_PROCESSING",
    "event_content": {
      "content": {
        "RefundsResponse": {
          "merchant_refund_id": "ref1782817526",
          "connector_refund_id": "refund_SoDJwDyhCtW5OmAM9WUzP",
          "status": "REFUND_PENDING",
          "error": {
            "connector_details": {}
          },
          "status_code": 200,
          "response_headers": {},
          "merchant_transaction_id": "azharamin-JUSP1782817439-1"
        }
      }
    },
    "source_verified": true,
    "event_ack_response": {
      "status_code": 200,
      "headers": {},
      "body": []
    },
    "supported_integrity_checks": []
  }
```

### CODE REVIEW

**Comment by XyneSpaces** [comment]
[nit] **Payment method additions:** The PPRO connector additions for Afterpay and UPI Intent/QR look good. A few verification items:

1. **Afterpay requirements:** Confirm that consumer email and billing country validation is enforced in the request validation layer—PPRO rejects Afterpay charges without these fields.

2. **UPI QR mapping:** The `UpiQr` → `ScanCode` and `UpiIntent` → `AppIntent` mapping logic should include test cases to verify the correct redirect URL construction.

3. **Webhook handling:** The new `WEBHOOK_REFUND_PROCESSING` event (enum 42) should be documented in the webhook integration guide. Verify that `PaymentChargeRefundPending` from PPRO correctly transitions the refund to `processing` state without duplicate events.

4. **Auth redirect priority:** The `build_auth_redirect` helper with `AppIntent → ScanCode → Redirect` priority should have a comment explaining this precedence logic for future maintainers.

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 1]
```diff
@@ -194,15 +189,44 @@ where
...
+                .resource_common_data
+                .get_billing_address()
+                .ok()
+                .and_then(|billing| billing.country);
+            if billing_country.is_none() {
+                return Err(IntegrationError::MissingRequiredField {
+                    field_name: "billing.address.country",
+                    context: Default::default(),
+                }
+                .into());
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs]: Lets directly throw error from get_billing_address() here

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 2]
```diff
@@ -83,11 +84,10 @@ pub struct PproAuthenticationSettings {
 pub struct PproAuthSettingsDetails {
     #[serde(skip_serializing_if = "Option::is_none")]
     pub return_url: Option<String>,
-    // TODO: Uncomment when adding support for other authentication flows
-    // #[serde(skip_serializing_if = "Option::is_none")]
-    // pub scan_by: Option<String>,
-    // #[serde(skip_serializing_if = "Option::is_none")]
-    // pub mobile_intent_uri: Option<String>
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub scan_by: Option<String>,
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs:91]: Is scan_by a required field? We are currently passing None for it across all other places as well

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 3]
```diff
@@ -1030,6 +1043,71 @@ pub struct PproAgreementRequest {
...
+    fn ppro_payment_method_type(&self) -> Option<common_enums::PaymentMethodType> {
+        None
+    }
+}
+
+impl GetPproPaymentMethodType for PaymentVoidData {
+    fn ppro_payment_method_type(&self) -> Option<common_enums::PaymentMethodType> {
+        None
+    }
+}
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs]: Is there a common trait we can reuse here. If not lets add a trait in the utils

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 4]
```diff
@@ -1030,6 +1043,71 @@ pub struct PproAgreementRequest {
...
+
+/// Builds a `RedirectForm` from the PPRO authentication method matching the requested
+/// payment method: UPI Intent → `APP_INTENT`, UPI QR → `SCAN_CODE`, otherwise `REDIRECT`.
+pub(crate) fn build_auth_redirect(
+    auth_methods: &[PproAuthenticationResponse],
+    payment_method_type: Option<common_enums::PaymentMethodType>,
+) -> Option<domain_types::router_response_types::RedirectForm> {
+    use domain_types::router_response_types::RedirectForm;
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs]: Lets avoid imports within function

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs (hunk 2 above)
Comment (ankush370) [reply crates/integrations/connector-integration/src/connectors/ppro/transformers.rs:91]: It's an optional field, it is used for setting expiry in case of UPI QR Transaction

**Review by AmitsinghTanwar007 — DISMISSED** [review]

**Review by JeevaRamu0104 — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 5]
```diff
@@ -1030,6 +1033,33 @@ pub struct PproAgreementRequest {
     pub instrument: Option<PproInstrument>,
 }
 
+/// Builds a `RedirectForm` from the PPRO authentication method matching the requested
+/// payment method: UPI Intent → `APP_INTENT`, UPI QR → `SCAN_CODE`, otherwise `REDIRECT`.
+pub(crate) fn build_auth_redirect(
+    auth_methods: &[PproAuthenticationResponse],
+    payment_method_type: Option<common_enums::PaymentMethodType>,
+) -> Option<RedirectForm> {
+    let wanted_type = match payment_method_type {
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs]: Please rename the variable to `wanted_type` to `matched_type`

File: crates/integrations/connector-integration/src/connectors/ppro.rs [hunk 6]
```diff
@@ -314,16 +314,12 @@ impl<T: PaymentMethodDataTypes + Debug + Sync + Send + 'static + Serialize>
...
 
-        let refund_id = charge
-            .refund_id
-            .clone()
-            .unwrap_or_else(|| charge.payment_charge_id.clone());
-
         Ok(
             domain_types::connector_types::RefundWebhookDetailsResponse {
-                connector_refund_id: Some(refund_id.clone()),
+                connector_refund_id: charge.refund_id.clone(),
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/ppro.rs:319]: **[S2] Dropped the `payment_charge_id` fallback for `connector_refund_id`.** Previously this was `refund_id.unwrap_or_else(|| payment_charge_id)`; now `charge.refund_id.clone()` can be `None`. Downstream, `ForeignTryFrom<RefundWebhookDetailsResponse> for RefundResponse` does `connector_refund_id.unwrap_or_default()`, so a refund webhook without `refundId` now yields an **empty** `connector_refund_id` where it previously fell back to `payment_charge_id`.

Low risk (all observed PPRO refund webhooks include `refundId`), but it's a behavioral change with no test for the `None` case. Consider a unit test asserting the intended behavior when `refundId` is absent, or restore a documented fallback if empty downstream is undesirable.

File: crates/types-traits/domain_types/src/types.rs [hunk 7]
```diff
@@ -8110,6 +8110,7 @@ pub fn generate_refund_sync_response(
                 connector_refund_id: response.connector_refund_id.clone(),
                 status: grpc_status as i32,
                 merchant_refund_id: Some(response.connector_refund_id.clone()),
+                merchant_transaction_id: None,
```
Comment (JeevaRamu0104) [inline crates/types-traits/domain_types/src/types.rs:8115]: **[S3] `merchant_transaction_id` is only ever populated on the webhook path.** Here (`generate_refund_sync_response`) and in `generate_refund_response` it's hardcoded `None`, so a merchant reconciling via the direct Refund or RSync response — rather than the webhook — never receives it, even for PPRO. This matches the stated design ("all other connectors set it to None"), but the new proto field's surface is narrower than "refund response" suggests. Worth a code comment or follow-up.

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 8]
```diff
@@ -194,15 +190,33 @@ where
...
 
+        // PPRO requires consumer email and country for Afterpay payments.
+        if matches!(
+            router_data.request.payment_method_type,
+            Some(common_enums::PaymentMethodType::AfterpayClearpay)
+        ) {
+            if email.is_none() {
+                return Err(IntegrationError::MissingRequiredField {
+                    field_name: "email",
+                    context: Default::default(),
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs]: **[S3] Asymmetric / empty error context on Afterpay validation.** Missing email returns `MissingRequiredField { field_name: "email", context: Default::default() }`, while missing country flows through `get_billing_country()?`. The empty `context` matches the pre-existing PPRO convention in this file (so it's *not* a new anti-pattern), but enriching it — e.g. `additional_context: "Afterpay requires consumer email"` — would improve parity with the country path and the newer-connector convention. Nit.

File: crates/integrations/connector-integration/src/connectors/ppro/transformers.rs [hunk 9]
```diff
@@ -1030,6 +1033,33 @@ pub struct PproAgreementRequest {
...
+
+    auth_methods.iter().find_map(|method| {
+        if method.r#type != wanted_type {
+            return None;
+        }
+        let details = method.details.as_ref()?;
+        let uri = match wanted_type {
+            PproAuthenticationType::AppIntent => details.mobile_intent_uri.clone(),
+            PproAuthenticationType::ScanCode => details.code_payload.clone(),
+            PproAuthenticationType::Redirect => details.request_url.clone(),
```
Comment (JeevaRamu0104) [inline crates/integrations/connector-integration/src/connectors/ppro/transformers.rs:1061]: **[S3] Unreachable match arm.** `wanted_type` is only ever `AppIntent` / `ScanCode` / `Redirect`, so the `_ => None` arm below this match is dead code. Harmless, but removable for clarity.

**Review by JeevaRamu0104 — COMMENTED** [review]
Non-blocking findings (no S0/S1 blockers). Proto field numbering is safe (42 unused, 24 appended), and all 14 `RefundWebhookDetailsResponse` constructors + both `EventType`↔grpc directions are in sync. Inline notes below.

File: crates/integrations/connector-integration/src/connectors/ppro.rs (hunk 6 above)
Comment (ankush370) [reply crates/integrations/connector-integration/src/connectors/ppro.rs:319]: Payment Charge Id is connector's transaction ID. It should not be mapped with connector_refund_id

File: crates/types-traits/domain_types/src/types.rs (hunk 7 above)
Comment (ankush370) [reply crates/types-traits/domain_types/src/types.rs:8115]: We only get this field in webhooks not in refund and refund sync. Added comment for clarity

**Review by JeevaRamu0104 — APPROVED** [review]

**Review by AmitsinghTanwar007 — APPROVED** [review]
